File: go/501.go
package main

import "math"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var ret []int
var maxFreq, curVal, curFreq int

func helper(cur *TreeNode)  {
  if cur == nil {
    return
  }
  helper(cur.Left)

  if cur.Val == curVal {
    curFreq++
  } else {
    curVal = cur.Val
    curFreq = 1
  }

  if curFreq == maxFreq {
    ret = append(ret, curVal)
  } else if curFreq > maxFreq {
    ret = nil
    ret = append(ret, curVal)
    maxFreq = curFreq
  }

  helper(cur.Right)
}

func findMode(root *TreeNode) []int {
	ret = nil
	maxFreq, curVal, curFreq = 0, math.MinInt32, 0
	helper(root)
	return ret
}

File: cpp/501.cpp
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  vector<int> ret;
  int maxFreq = 0, curFreq = 0, preVal = INT_MIN;
  void helper(TreeNode *cur) {
    if (cur == nullptr)
      return;
    helper(cur->left);

    if (cur->val == preVal)
      curFreq++;
    else {
      preVal = cur->val;
      curFreq = 1;
    }

    if (curFreq > maxFreq) {
      ret.clear();
      ret.push_back(cur->val);
      maxFreq = curFreq;
    } else if (curFreq == maxFreq) {
      ret.push_back(cur->val);
    }

    helper(cur->right);
  }

  vector<int> findMode(TreeNode *root) {
    helper(root);
    return ret;
  }
};
